### PR TITLE
[SPARK-10429] [SQL] make mutableProjection atomic

### PR DESCRIPTION
**Pull request by davies**
Right now, SQL's mutable projection updates every value of the mutable project after it evaluates the corresponding expression. This makes the behavior of MutableProjection confusing and complicate the implementation of common aggregate functions like stddev because developers need to be aware that when evaluating {{i+1}}th expression of a mutable projection, {{i}}th slot of the mutable row has already been updated.

This PR make the MutableProjection atomic, by generating all the results of expressions first, then copy them into mutableRow.

Had run a mircro-benchmark, there is no notable performance difference between using class members and local variables. 

cc @yhuai

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Projection.scala [hunk 1]
```diff
@@ -79,7 +80,13 @@ case class InterpretedMutableProjection(expressions: Seq[Expression]) extends Mu
...
+      // Store the result into buffer first, to make the projection atomic (needed by aggregation)
+      buffer(i) = exprArray(i).eval(input)
+      i += 1
+    }
+    i = 0
+    while (i < exprArray.length) {
+      mutableRow(i) = buffer(i)
       i += 1
     }
     mutableRow
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Projection.scala]: should we assign null to `buffer` at the end?

**Comment by cloud-fan** [comment]
LGTM, pending test

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala [hunk 2]
```diff
@@ -371,101 +371,77 @@ abstract class StddevAgg(child: Expression) extends DeclarativeAggregate {
...
   private val resultType = DoubleType
 
-  private val preCount = AttributeReference("preCount", resultType)()
-  private val currentCount = AttributeReference("currentCount", resultType)()
-  private val preAvg = AttributeReference("preAvg", resultType)()
-  private val currentAvg = AttributeReference("currentAvg", resultType)()
-  private val currentMk = AttributeReference("currentMk", resultType)()
+  private val count = AttributeReference("count", resultType)()
+  private val avg = AttributeReference("avg", resultType)()
+  private val mk = AttributeReference("mk", resultType)()
```
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala]: @JihongMA Can you take a look at `Stddev`'s change? We are changing the way that we evaluate expressions in mutable projection. We will first evaluate all expressions and then update the mutable row.

**Comment by marmbrus** [comment]
Thanks, I'm going to go ahead and merge this.  @JihongMA , please let us know if there is anything that needs to be addressed in a follow-up.

**Comment by JihongMA** [comment]
@yhuai and @marmbrus, thank you for checking with me. I have a pending PR (SPARK-11420)to change stddev support through ImerativeAggregate. so please go ahead and I will take care stddev support.

**Comment by davies** [comment]
For stddev/variance, it's clear that using ImerativeAggregate will be faster. Then we don't have case that need the MutableProject to be atomic, should we revert this PR? cc @marmbrus @yhuai

**Comment by yhuai** [comment]
ah sorry just saw this message. Since the changed behavior is easier to reason about, let's keep it.

**Pull request closed without merging**